packages/hyperliquid-composer/contracts/extensions/PreFundedFeeAbstraction.sol

```diff
@@ -0,0 +1,245 @@
+                uint256 requiredCoreBalance = maxUsersPerBlock * QUOTE_ASSET_DECIMALS;
+                if (coreBalance < requiredCoreBalance) revert InsufficientCoreBalance(coreBalance, requiredCoreBalance);
+
+                uint64 feeCollected = originalAmount - coreAmount;
+                emit FeeCollected(_to, feeCollected);
+            }
+
+            IERC20(ERC20).safeTransfer(ERC20_ASSET_BRIDGE, amounts.evm);
+            _submitCoreWriterTransfer(_to, ERC20_CORE_INDEX_ID, coreAmount);
+        } else {
+            if (msg.value > 0 && shouldActivate) revert HYPEActivationNotAllowed();
+        }
+    }
+
+    /**
+     * @notice Calculates activation fee in asset tokens using current spot price.
+     * @return Activation fee amount in core asset decimals
+     */
+    function activationFee() public view virtual override returns (uint64) {
+        uint64 rawPrice = _spotPx(SPOT_PAIR_ID);
+
+        /// @dev Prevent zero-fee edge case: When rawPrice > ACTIVATION_FEE_NUMERATOR,
+        ///      the division ACTIVATION_FEE_NUMERATOR / rawPrice < 1 rounds down to 0.
+        ///      This would cause silent failures where Core expects payment but contract calculates zero fee.
+        /// @dev Trigger thresholds (ACTIVATION_COST ≈ $1.50):
+        ///      - szDecimals = 0 (min): token_price > $1B USD
+        ///      - szDecimals = 5 (max): token_price > $150M USD (ex: BTC)
+        if (rawPrice > ACTIVATION_FEE_NUMERATOR) revert PriceExceedsActivationFeeNumerator(rawPrice);
+
+        return uint64(ACTIVATION_FEE_NUMERATOR / rawPrice);
+    }
+
+    /**
+     * @notice Retrieves quote tokens from HyperCore to a specified address
+     * @dev Transfers quote tokens from the composer's HyperCore balance to the specified address
+     * @dev Can only be called by the recovery address
+     * @param _coreAmount Amount of quote tokens to retrieve in HyperCore decimals, or FULL_TRANSFER for all
+     * @param _to Destination address to receive the retrieved quote tokens
+     */
+    function retrieveQuoteTokens(uint64 _coreAmount, address _to) public virtual onlyRecoveryAddress {
+        /// @dev Balance reads are asynchronous: spotBalance() returns the state from the previous HyperCore block.
+        ///      If activation txs executed earlier in this block (ex: tx 1-4) and this withdrawal is later (ex: tx 5),
+        ///      those activations consumed quote tokens but the balance reduction won't be visible until next block.
+        ///      Withdrawing the full visible balance may fail on Core since actual balance is already lower.
+        uint64 maxTransferAmt = _getMaxTransferAmount(QUOTE_ASSET_INDEX, _coreAmount);
+
+        _submitCoreWriterTransfer(_to, QUOTE_ASSET_INDEX, maxTransferAmt);
+        feeWithdrawalBlockNumber = block.number;
+        emit Retrieved(QUOTE_ASSET_INDEX, maxTransferAmt, _to);
+    }
```

Comment:
Should the concept of quote token (instead of USDC) and this function be in `RecoverableComposer`, and then apply the `feeWithdrawalBlockNumber = block.number` line in an override here?